custom_components/tesla_custom/base.py

```diff
@@ -0,0 +1,178 @@
+    ) -> None:
+        """Initialise the Tesla device."""
+        super().__init__(coordinator)
+        self._coordinator = coordinator
+        self._enabled_by_default: bool = True
+        self.attrs: dict[str, str] = {}
+        self.hass = hass
+        self.type = None
+
+    def refresh(self) -> None:
+        """Refresh the device data.
+
+        This is called by the DataUpdateCoodinator when new data is available.
+
+        This assumes the controller has already been updated. This should be
+        called by inherited classes so the overall device information is updated.
+        """
+        self.async_write_ha_state()
+
+    @property
+    def name(self) -> str:
+        """Return device name."""
+        return self.type.capitalize()
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Set entity registry to default."""
+        return self._enabled_by_default
+
+    async def async_added_to_hass(self) -> None:
+        """Register state update callback."""
+        self.async_on_remove(self.coordinator.async_add_listener(self.refresh))
+
+
+class TeslaCarEntity(TeslaBaseEntity):
+    """Representation of a Tesla car device."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        car: TeslaCar,
+        coordinator: TeslaDataUpdateCoordinator,
+    ) -> None:
+        """Initialise the Tesla car device."""
+        super().__init__(hass, coordinator)
+        self._car = car
+
+    async def update_controller(
+        self, *, wake_if_asleep: bool = False, force: bool = True, blocking: bool = True
+    ) -> None:
+        """Get the latest data from Tesla.
+
+        This does a controller update then a coordinator update.
+        The coordinator triggers a call to the refresh function.
+
+        Setting the blocking param to False will create a background task for the update.
+        """
+
+        if blocking is False:
+            await self.hass.async_create_task(
+                self.update_controller(wake_if_asleep=wake_if_asleep, force=force)
+            )
+            return
+
+        await self._coordinator.controller.update(
+            self._car.id, wake_if_asleep=wake_if_asleep, force=force
+        )
+        await self._coordinator.async_refresh()
+
+    @property
+    def available(self) -> bool:
+        """Return availability of data."""
+        return self._car.data_available != {}
+
+    @property
+    def vehicle_name(self) -> str:
+        """Return vehicle name."""
+        return (
+            self._car.display_name
+            if self._car.display_name is not None
+            and self._car.display_name != self._car.vin[-6:]
+            else f"Tesla Model {str(self._car.vin[3]).upper()}"
+        )
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for car entity."""
+        return slugify(
+            f"Tesla Model {str(self._car.vin[3]).upper()} {self._car.vin[-6:]} {self.type}"
```

Comment:
Pretty minor but Tesla is obviously coming out with vehicles other than "Model ___" like the Cybertruck.. not sure why we even need the vehicle name in the unique_id since this isn't readily visible to the user. Since we're already going to have a decent amount of breaking changes, I'm thinking to just change this to something more generic but that will.

Reply:
Ah makes sense. Since we're going to have a bunch of breaking changes anyways, I'm just setting this to VIN + type.